ripe/atlas/sagan/ssl.py

```diff
@@ -58,32 +55,59 @@ def __init__(self, data, **kwargs):
 
         self.has_expired = None
 
-        # Clean up the certificate data and use OpenSSL to parse it
-        x509 = OpenSSL.crypto.load_certificate(
-            OpenSSL.crypto.FILETYPE_PEM,
-            data.replace("\\/", "/").replace("\n\n", "\n")
-        )
-        subject = dict(x509.get_subject().get_components())
-        issuer = dict(x509.get_issuer().get_components())
+        self.extensions = {}
+
+        cert = x509.load_pem_x509_certificate(data.encode('ascii'), openssl.backend)
+
+        if cert:
+            self.checksum_md5 = self._colonify(cert.fingerprint(hashes.MD5()))
+            self.checksum_sha1 = self._colonify(cert.fingerprint(hashes.SHA1()))
+            self.checksum_sha256 = self._colonify(cert.fingerprint(hashes.SHA256()))
+
+            self.valid_from = pytz.utc.localize(cert.not_valid_before)
+            self.valid_until = pytz.utc.localize(cert.not_valid_after)
 
-        if x509 and subject and issuer:
+            self.has_expired = self._has_expired()
 
-            self.subject_cn = self._string_from_dict_or_none(subject, b"CN")
-            self.subject_o = self._string_from_dict_or_none(subject, b"O")
-            self.subject_c = self._string_from_dict_or_none(subject, b"C")
-            self.issuer_cn = self._string_from_dict_or_none(issuer, b"CN")
-            self.issuer_o = self._string_from_dict_or_none(issuer, b"O")
-            self.issuer_c = self._string_from_dict_or_none(issuer, b"C")
+            self._add_extensions(cert)
 
-            self.checksum_md5 = x509.digest("md5").decode()
-            self.checksum_sha1 = x509.digest("sha1").decode()
-            self.checksum_sha256 = x509.digest("sha256").decode()
+        if cert and cert.subject:
+            self.subject_cn, self.subject_o, self.subject_c = \
+                self._parse_x509_name(cert.subject)
 
-            self.has_expired = bool(x509.has_expired())
+        if cert and cert.issuer:
+            self.issuer_cn, self.issuer_o, self.issuer_c = \
+                self._parse_x509_name(cert.issuer)
 
-            self.valid_from = None
-            self.valid_until = None
-            self._process_validation_times(x509)
+    def _add_extensions(self, cert):
+        for ext in cert.extensions:
+            if ext.oid._name == 'subjectAltName':
+                self.extensions['subjectAltName'] = []
+                for san in ext.value:
+                    self.extensions['subjectAltName'].append(san.value)
+
+    @staticmethod
+    def _colonify(bytes):
+        hex = codecs.getencoder('hex_codec')(bytes)[0].decode('ascii').upper()
+        return ':'.join(a+b for a,b in zip(hex[::2], hex[1::2]))
+
+    @staticmethod
+    def _parse_x509_name(name):
+        cn = None
+        o = None
+        c = None
+        for attr in name:
+            if attr.oid.dotted_string == '2.5.4.6': # country
```

Comment:
Arbitrary strings like `"2.5.4.6"` don't make a lot of sense out of context like this, so it might be nicer to place them in a class constant like `OLD_STYLE_COUNTRY="2.5.4.6"` (or a more appropriate name).  I don't really get what's going on here, but maybe that's just my ignorance of arcane SSL rules.  If it's particularly weird though, a docstring on the function might also be nice.